### ISSUE TITLE
Visualising relative abundance of sample population from two marker genes in one barplot 

### ISSUE THREAD

**Issue by cathreenj** [joey711/phyloseq#1221]
Hi, 
I don't know if this question is redundant but my first search on this topic was unsuccessfull and I'm still a beginner in this field.
I have microbiome data from two marker genes 16S and a functional gene specific for a certain clade. Now I would like to compare the results for the same clade but from different marker genes in one barplot. I read the manual for the merge_phyloseq command, but I'm not sure this will work, since the data come from two different analyses and two different reference databases were used to assign the taxonomies. 
Is there another away to plot the data in the same barplot? 
I'd be very glad if you could point me in the right direction.
Thank you 
Cathrine

**Comment by mikemc** [comment]
What are the taxonomy systems for each phyloseq object? If the taxonomy names are not compatible, e.g. if one phyloseq object is using NCBI taxonomy identifiers and the other is using Silva or RDP identifiers, then there will not be a straightforward way to proceed.

Do you have the reference sequences for the taxa available? E.g., if the data were analyzed with DADA2, so that you have the ASV sequences for each taxon, then the easiest thing to do would be to redo the taxonomy assignment from each against a single database before trying to proceed.

**Comment by cathreenj** [comment]
Hey, thank you so much for the fast answer!

For the 16S analysis I used the silva_nr_v132_train_set.fa file and for the mcra analysis, I used a reference database, which I reformatted myself to be compatible with DADA2.
Yes, both datasets were analysed with DADA2.

So you would suggest, that I merge the two reference databases eventhough they refer to different genes?

**Comment by mikemc** [comment]
No, don't merge the databases; you do want different databases for each gene. But they should use the same taxonomy (e.g., both should use Silva taxonomic identifiers). Silva taxonomy is different from and incompatible with Greengenes, NCBI, RDP, etc. (E.g., the genera literally have different names, and don't necessarily correspond to each other). If you assigned Silva taxonomies to both genes, then you can merge the phyloseq objects with `merge_phyloseq()` and then aggregate to  to e.g. genus or family with `tax_glom()`, and the ASVs with the same genus will be merged into one OTU (named after the most abundant of the merged ASVs).

I suggest making sure the `taxa_names()` are different in the two phyloseq objects before you call `merge_phyloseq()`. You might want to use different sample names---which will keep the data from each marker gene separate---or the sample sample names, which will merge the data from each sample into one row in the `otu_table`. Try experimenting with a simple example first, as in `help(merge_phyloseq)`. Here is a simplified e.g. to show the behavior of `merge_phyloseq()` when the taxa are differently named but the samples have the same names.
```r
x  <- otu_table(matrix(sample(0:5,100,TRUE),10,10), taxa_are_rows=FALSE)
y <- x
taxa_names(y) <- paste0("sp", 11:20)
xy <- merge_phyloseq(x, y)
```
I would make sure you understand the merging and aggregating of tax_glom with a small dataset, e.g., the first 5 samples and 5 taxa from each of your phyloseq objects, to make sure you understand what you end up with.

**Comment by mikemc** [comment]
The easiest thing to do, which requires no special steps, is to simply create bar plots for each marker-gene phyloseq object separately, with the `fill` parameter set to the same taxonomic rank (e.g., `fill = "Phylum"), and compare the two plots. If you want to plot the two bar plots side by side, I suggest the `plot_grid()` function from the `cowplot` package.

**Comment by cathreenj** [comment]
Hey, thank you so much for your help!
Up to now that was my first approach I just plotted the two communities separately but with the same taxonomic rank. 
I think I will check out the cow package and plot_grid function. 
I really don't think it will be worth the effort to harmonise the taxonomies of all my entries in the two reference databases.
Ok, so basically I have microbiome data from two sampling points that were analysed using two marker genes. I would like to create one plot containing the communities of each sampling point resulting from the two respective marker genes as stacked barcharts side by side. I want to show that the results from the two analyses coincide.

**Comment by mikemc** [comment]
Here's an illustration of one way to go about this. Some quick background: The `plot_bar()` function first called the `psmelt()` function to get a data frame in "tidy" or "melted" form, and then uses `ggplot2` to create the plot. To make customized bar plots you are often better off calling `psmelt()` (or use the `justDF=TRUE` option to `plot_bar()`) and then creating the ggplot yourself. In order to plot the data from both phyloseq objects in the same plot, you need to get data frames from each, and combine them, while adding a new column (I'll call "Marker") that tracks which rows are from which phyloseq object (and thus which marker gene). Once you have this data frame, you can use ggplot with "faceting" plot bar charts for each marker side-by-side, for each sample.

I'm going to use [speedyseq](https://github.com/mikemc/speedyseq) package in my demonstration to speed up `psmelt()`, but you can just replace "speedyseq" with "phyloseq".

For example data, I'll use the first 8 samples of the GlobalPatterns dataset. I'm just going to use identical copies for the two different marker genes, but in your case you'd use the two different phyloseq objects for `ps1` and `ps2`.
```r
library(speedyseq)
data(GlobalPatterns)
ps <- GlobalPatterns
ps0 <- prune_samples(sample_names(ps)[1:8], ps)
ps1 <- ps0
ps2 <- ps0
```
Now, supposing `ps1` and `ps2` are the data for the two different marker genes, and have matching sample names, you can do
```r
ps1.prop <- transform_sample_counts(ps1, function(x) x / sum(x))
ps2.prop <- transform_sample_counts(ps2, function(x) x / sum(x))
tb1 <- psmelt(ps1.prop)
tb2 <- psmelt(ps2.prop)
tb <- bind_rows(list(Marker1 = tb1, Marker2 = tb2), .id = "Marker")
ggplot(tb.top, aes(Marker, Abundance, fill = Phylum)) +
    geom_col() +
    facet_wrap(~Sample)
```
This plot is overcrowded here because there are so many phyla. In reality you probably want to subset to the most abundant ~10 phyla or something, and I'm guessing there are other issues showing how to do that.
